# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaManagerImpl.h
# Commit: b84880c5de44
# Full Hash: b84880c5de443e6274953d210444c57319ce4b5f
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-13 09:18:36
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Make it possible to initialize temporary origins without ensuring origin directories; r=dom-storage-reviewers,jari
#   
#   LSNG already uses some QuotaManager APIs to achieve that origin directories are
#   not created if they don't exist during datastore preparation, but the feature
#   is not easy to use and it's also not generalized enough for use in other quota
# ==============================================================================

diff -r c5f28ed5382b -r b84880c5de44 dom/quota/QuotaManagerImpl.h
--- a/dom/quota/QuotaManagerImpl.h	Sat Oct 12 23:42:06 2024 +0300
+++ b/dom/quota/QuotaManagerImpl.h	Sat Oct 12 20:58:36 2024 +0000
@@ -14,6 +14,19 @@
 
 namespace mozilla::dom::quota {
 
+template <typename F>
+auto QuotaManager::WithOriginInfo(const OriginMetadata& aOriginMetadata,
+                                  F aFunction)
+    -> std::invoke_result_t<F, const RefPtr<OriginInfo>&> {
+  MutexAutoLock lock(mQuotaMutex);
+
+  RefPtr<OriginInfo> originInfo =
+      LockedGetOriginInfo(aOriginMetadata.mPersistenceType, aOriginMetadata);
+  MOZ_ASSERT(originInfo);
+
+  return aFunction(originInfo);
+}
+
 template <typename P>
 void QuotaManager::CollectPendingOriginsForListing(P aPredicate) {
   MutexAutoLock lock(mQuotaMutex);
@@ -28,7 +41,7 @@
         pair->LockedGetGroupInfo(PERSISTENCE_TYPE_DEFAULT);
     if (groupInfo) {
       for (const auto& originInfo : groupInfo->mOriginInfos) {
-        if (!originInfo->mDirectoryExists) {
+        if (!originInfo->LockedDirectoryExists()) {
           aPredicate(originInfo);
         }
       }